pkg/etcd/transport/client.go

```diff
@@ -0,0 +1,169 @@
+package transport
+
+import (
+	"context"
+	"fmt"
+	"path"
+	"strings"
+
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"go.uber.org/fx"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/types/known/anypb"
+
+	etcdclient "github.com/fluxninja/aperture/v2/pkg/etcd/client"
+	"github.com/fluxninja/aperture/v2/pkg/log"
+)
+
+// TransportClientModule is the client fx provider for etcd transport.
+var TransportClientModule = fx.Options(
+	fx.Provide(NewEtcdTransportClient),
+)
+
+// EtcdTransportClient is the client side for the etcd transport.
+type EtcdTransportClient struct {
+	etcdClient *etcdclient.Client
+	Registry   *HandlerRegistry
+}
+
+// NewEtcdTransportClient creates and returns a new etcd transport client module.
+func NewEtcdTransportClient(client *etcdclient.Client) *EtcdTransportClient {
+	return &EtcdTransportClient{
+		etcdClient: client,
+		Registry:   NewHandlerRegistry(),
+	}
+}
+
+// HandlerRegistry allow registering handlers and can start a dispatcher.
+//
+// This is intended to be used at at fx provide/invoke stage.
+type HandlerRegistry struct {
+	handlers map[protoreflect.FullName]untypedHandler
+}
+
+type untypedHandler func(context.Context, *anypb.Any) (proto.Message, error)
+
+// NewHandlerRegistry creates a new HandlerRegistry.
+func NewHandlerRegistry() *HandlerRegistry {
+	return &HandlerRegistry{
+		handlers: map[protoreflect.FullName]untypedHandler{},
+	}
+}
+
+// RegisterWatcher allows to register a client on the etcd transport.
+func RegisterWatcher(lc fx.Lifecycle, t *EtcdTransportClient, agentName string) {
+	lc.Append(fx.Hook{
+		OnStart: func(ctx context.Context) error {
+			go t.RegisterWatcher(agentName)
+			return nil
+		},
+		OnStop: func(ctx context.Context) error {
+			return nil
+		},
+	})
+}
+
+// RegisterWatcher register an agent on the etcd transport client.
+func (c *EtcdTransportClient) RegisterWatcher(agentName string) {
+	path := path.Join(RPCBasePath, RPCRequestPath, agentName)
+	watchCh := c.etcdClient.Watch(context.Background(), path, clientv3.WithPrefix())
+	for watchResp := range watchCh {
```

Comment:
Need to restart the watch on non-recoverable errors. Check out and try to re-use [this implementation](https://github.com/fluxninja/aperture/blob/main/pkg/etcd/watcher/watcher.go#L128)

Reply:
this function is now being called in a loop with errors handled